### PR TITLE
Make edxapp pipeline system configuration atomic

### PR DESCRIPTION
**Pull request by cpennington**

### CODE REVIEW

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+                '-c',
+                'echo Manual Verification run number $GO_PIPELINE_COUNTER completed by $GO_TRIGGER_USER'
+            ],
+        )
+    )
+
+    return pipeline
+
+
+def rollback_asgs(edxapp_deploy_group, pipeline_name, rollback_pipeline, variable_files, cmd_line_vars):
```
Comment (macdiesel) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Need better doc strings here. Is the rollback_pipeline the pipeline that you are fetching the deploy phase from? (just checked and yes). Perhaps call this last_deploy_pipeline? When reading the code I initially thought this was this pipeline (the rollback pipeline).

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+    Variables needed for this pipeline:
+    materials: List of dictionaries of the materials used in this pipeline
+    upstream_pipelines: List of dictionaries of the upstream piplines that feed in to the rollback pipeline.
+    """
+    config = utils.merge_files_and_dicts(variable_files, list(cmd_line_vars,))
+
+    pipeline = edxapp_deploy_group.ensure_replacement_of_pipeline(pipeline_name)\
+                                  .ensure_environment_variables({'WAIT_SLEEP_TIME': config['tubular_sleep_wait_time']})
+
+    for material in config['materials']:
```
Comment (macdiesel) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Should this use the materials.py file?

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+            'JENKINS_JOB_TOKEN': config['jenkins_job_token']
+        }
+    )
+
+    # Create the stage with the Jenkins jobs
+    jenkins_stage = pipeline.ensure_stage(constants.JENKINS_VERIFICATION_STAGE_NAME)
+    jenkins_stage.set_has_manual_approval()
+    jenkins_user_name = config['jenkins_user_name']
+
+    jenkins_url = "https://test-jenkins.testeng.edx.org"
```
Comment (macdiesel) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Should this be parameterized? It's just weird to me that we parameterize the username and password but then hard code the url.

File: edxpipelines/utils.py [hunk 4]
```diff
@@ -110,7 +110,7 @@ def merge_files_and_dicts(file_paths, dicts):
         else:
             raise ValueError("dicts contains an instance that is not a dictionary {}".format(d.__class__))
 
-    [file_variables.append(d) for d in dict_vars]
+    file_variables.extend(dict_vars)
```
Comment (macdiesel) [inline edxpipelines/utils.py]: Thank you python!

File: tox.ini [hunk 5]
```diff
@@ -4,6 +4,6 @@ skip_install = True
 skipsdist = True
 
 [testenv]
-deps = -rrequirements.txt
-commands = {posargs:python deploy_pipelines.py --dry-run -v tools -f config.yml}
+deps = -rrequirements/test_requirements.txt
```
Comment (macdiesel) [inline tox.ini]: is -rrequirements correct here or should there be a space?

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+        config['play_name'],
+        "edx",
+        "stage",
+        base_ami_id
+    )
+
+    return pipeline
+
+
+def build_migrate_deploy_subset_pipeline(
```
Comment (cpennington) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Yes, that's coming in a later PR.

File: edxpipelines/pipelines/edxapp_pipelines.py (hunk 3 above)
Comment (cpennington) [inline edxpipelines/pipelines/edxapp_pipelines.py]: It's public, unlike the username and password. In the future, yes, I'm open to more parameterization.

File: edxpipelines/pipelines/edxapp_pipelines.py (hunk 1 above)
Comment (cpennington) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Will do.

File: tox.ini (hunk 5 above)
Comment (cpennington) [inline tox.ini]: It's fine w/o a space (a space would probably work as well).

File: edxpipelines/pipelines/edxapp_pipelines.py (hunk 1 above)
Comment (cpennington) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Added.

**Comment by cpennington** [comment]
@macdiesel: I've addressed all of your comments.

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 7]
```diff
@@ -416,9 +416,18 @@ def manual_verification(edxapp_deploy_group, variable_files, cmd_line_vars):
 
 def rollback_asgs(edxapp_deploy_group, pipeline_name, rollback_pipeline, variable_files, cmd_line_vars):
     """
-    Variables needed for this pipeline:
-    materials: List of dictionaries of the materials used in this pipeline
-    upstream_pipelines: List of dictionaries of the upstream piplines that feed in to the rollback pipeline.
+    Arguments:
+        edxapp_deploy_group (gomatic.PipelineGroup): The group in which to create this pipeline
+        pipeline_name (str): The name of this pipeline
+        rollback_pipeline (gomatic.Pipeline): The pipeline to retrieve the ami_deploy_info.yml artifact from
```
Comment (macdiesel) [inline edxpipelines/pipelines/edxapp_pipelines.py]: docstring good, could I talk you in to a new name? deploy_pipeline maybe? the naming is ambiguous to me

Comment (cpennington) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Sure, I'll change it.

**Comment by cpennington** [comment]
AFAIK, the sandbox environment doesn't exist. I'm relying on the diff of the gocd configuration, and manual inspection of such.

File: canonicalize_gocd.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+def canonicalize(child_sort_key=None, post_process=None):
+    """
+    Build a canonicalizer function using the specified key to sort
+    element children. (Will not sort children if child_sort_key is None).
+    """
+    def canonicalizer(element):
+        canon = copy(element)
+        canon.tail = None
+        canon[:] = [canonicalize_element(child) for child in canon]
```
Comment (doctoryes) [inline canonicalize_gocd.py]: I though `canon[:]` was "return a copy of canon"? What does this and the one below do?

File: edxpipelines/pipelines/cd_edxapp_latest.py [hunk 9]
```diff
@@ -74,238 +98,152 @@ def install_pipelines(save_config_locally, dry_run, bmd_steps, variable_files, c
...
+            TUBULAR, CONFIGURATION, EDX_PLATFORM, EDX_SECURE, EDGE_SECURE, MCKINSEY_SECURE,
+            EDX_MICROSITE, EDX_INTERNAL, EDGE_INTERNAL, MCKINSEY_INTERNAL
+        ):
+            pipeline.ensure_material(material)
+
+    manual_verification = edxapp_pipelines.manual_verification(edxapp_deploy_group, variable_files, cmd_line_vars)
+
+    rollback_edx = edxapp_pipelines.rollback_asgs(
+        edxapp_deploy_group,
+        'PROD_edx_edxapp_Rollback_latest',
```
Comment (doctoryes) [inline edxpipelines/pipelines/cd_edxapp_latest.py]: JFTR, I regret using `latest` - should have just used `v1`. But I realize you're keeping the names the same for now.

File: edxpipelines/pipelines/edxapp_pipelines.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+        )
+    }
+
+    pipeline = pipeline_group.ensure_replacement_of_pipeline(pipeline_name)
+
+    # We always need to launch the AMI, independent deploys are done with a different pipeline
+    # if the pipeline has a migrate stage, but no building stages, look up the build information from the
+    # upstream pipeline.
+    ami_artifact = None
+    if 'm' in bmd_steps and 'b' not in bmd_steps:
```
Comment (doctoryes) [inline edxpipelines/pipelines/edxapp_pipelines.py]: Yep - confuses me. And I wrote it! ☹️

File: canonicalize_gocd.py (hunk 8 above)
Comment (cpennington) [inline canonicalize_gocd.py]: The `[:]` syntax updates all of the children of `canon`.

File: edxpipelines/pipelines/cd_edxapp_latest.py (hunk 9 above)
Comment (cpennington) [inline edxpipelines/pipelines/cd_edxapp_latest.py]: Yes

**Pull request merged**